REGRESSION (UI-side compositing) changing between non-overlay and overlay scrollbars should adjust the size of the scrollable area

nmoucht opened the pull request:
#### 81451dabde97ba73cbed2abc4c4be64957aa20f6
<pre>
REGRESSION (UI-side compositing) changing between non-overlay and overlay scrollbars should adjust the size of the scrollable area
<a href="https://bugs.webkit.org/show_bug.cgi?id=263618">https://bugs.webkit.org/show_bug.cgi?id=263618</a>
<a href="https://rdar.apple.com/117507268">rdar://117507268</a>

Reviewed by Simon Fraser.

This patch fixes a couple bugs related to switching back and forth dynamically between
overlay and non-overlay scrollbars. First, setFrameScrollingNodeState() was not always
called when switching the default, so iterate through each page and call
LocalFrameView::overlayScrollbarDefaultDidChange to force a scrolling commit to plumb
across this state change. Second, when we get the state change on the ui-process side,
we call _updateAllScrollerImpPairsForNewRecommendedScrollerStyle, which notifies all the
scroller imp pairs in the process of the new style. On the web process side, we had code
to adjust the scrollbar thickness and force a relayout, which now doesn&apos;t work since there
are no NSScrollerImpPairs in the web process. Instead, replicate this code in
RemoteScrollbarsController and loop through the scrollable areas in the FrameView. This
takes care of updating scrollable areas with overflow, while
<a href="https://github.com/WebKit/WebKit/pull/19561">https://github.com/WebKit/WebKit/pull/19561</a> will take care of updating scrollable areas
without overflow. Also add some additional testing infrastructure to compare the size of
the scrollable area before and after toggling scrollbar style.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::usesOverlayScrollbars const):
(WebCore::LocalFrameView::overlayScrollbarDefaultDidChange):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setFrameScrollingNodeState):
* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::scrollingTreeNodeMarkScrollbarStyleChange):
(WebCore::ScrollingTree::scrollingTreeNodesWithScrollbarStyleChange):
(WebCore::ScrollingTree::clearScrollingTreeNodesWithScrollbarStyleChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::markHasScrollbarStyleChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(-[WebScrollerImpPairDelegateMac scrollerImpPair:updateScrollerStyleForNewRecommendedScrollerStyle:]):
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::updateScrollbarThickness):
* Source/WebCore/platform/Scrollbar.h:
* Source/WebCore/platform/ScrollbarsController.cpp:
(WebCore::ScrollbarsController::updateScrollbarsThickness):
* Source/WebCore/platform/ScrollbarsController.h:
(WebCore::ScrollbarsController::updateScrollerStyle):
* Source/WebCore/platform/mac/ScrollbarsControllerMac.h:
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingNodesHaveScrollbarStyleChange):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeMarkScrollbarStyleChange):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::macScrollbarTheme):
(WebKit::RemoteScrollbarsController::updateScrollerStyle):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.messages.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::scrollingTreeNodesHaveScrollbarsWithStyleChange):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::scrollerStylePreferenceChanged):

Canonical link: <a href="https://commits.webkit.org/276439@main">https://commits.webkit.org/276439@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f42aefcfdfba3a48727085a49d434acfcc773d63

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44645 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23727 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47104 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47298 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40650 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46947 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27765 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21119 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45221 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20791 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38428 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/17768 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/18241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39570 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2692 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40861 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39856 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48934 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19599 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16136 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43661 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20929 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38686 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/42410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9950 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20597 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
```diff
@@ -118,5 +119,25 @@
+
+void RemoteScrollbarsController::updateScrollerStyle()
+{
+    auto* macTheme = macScrollbarTheme();
```
Why do you need to cast to the Mac theme? The base class has a virtual `usesOverlayScrollbars()`.

nmoucht replied:
This was following the code in `ScrollbarsControllerMac` but I believe its to avoid doing this for MockScrollbarsTheme.

smfr replied:
Better to ask the theme if it uses overlay scrollbars then.

nmoucht replied:
Changed to use the base class.

smfr commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -1199,6 +1199,11 @@ bool shouldTransformObject(id object) const override
     static_cast<ScrollbarThemeMac&>(theme).preferencesChanged();
     
+    for (auto& page : m_pageMap.values()) {
+        if (auto* frameView = page->localMainFrameView())
```
Safe coding rules say that this should be `if (RefPtr frameView = page->localMainFrameView())`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -6460,6 +6460,25 @@ void LocalFrameView::invalidateScrollAnchoringElement()
 
+bool LocalFrameView::usesOverlayScrollbars() const
+{
+    return ScrollbarTheme::theme().usesOverlayScrollbars();
```
`ScrollbarTheme::theme()` returns a process-global, so why does `LocalFrameView` need to wrap the call?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -6460,6 +6460,25 @@ void LocalFrameView::invalidateScrollAnchoringElement()
+{
+    if (auto* scrollableAreas = this->scrollableAreas()) {
+        for (auto& scrollableArea : *scrollableAreas) {
+            if (!scrollableArea.scrollbarsController().shouldRegisterScrollbars())
```
Not immediately clear why something about registering scrollbars affects whether we should call `updateScrollerStyle`. This complexity should be hidden inside `ScrollableArea.scrollbarsController`

nmoucht replied:
Moved this code to `scrollbarsController`.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -6460,6 +6460,25 @@ void LocalFrameView::invalidateScrollAnchoringElement()
+    }
+    if (!scrollbarsController().shouldRegisterScrollbars())
+        scrollbarsController().updateScrollerStyle();
+    if (auto scrollingCoordinator = this->scrollingCoordinator())
```
I'd expect a scrollbar size change to trigger layout, which should then trigger `RenderLayerCompositor::updateScrollingNodeForScrollingRole`, so not sure why this is necessary.

nmoucht replied:
Right now that we trigger layout in `updateScrollerStyleForScrollableAreas` this is no longer necessary.

smfr commented on Source/WebCore/page/LocalFrameView.h:
```diff
@@ -745,6 +745,9 @@ class LocalFrameView final : public FrameView {
     void invalidateScrollAnchoringElement() final;
 
+    bool usesOverlayScrollbars() const;
+    WEBCORE_EXPORT void overlayScrollbarDefaultDidChange();
```
It's a bit odd to see the word "Default" here since that's a term of art for Cocoa, but this is a cross-platform file. Maybe `scrollbarsStyleDidChange`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
```diff
@@ -137,6 +137,7 @@ class WEBCORE_EXPORT ScrollingTreeScrollingNode : public ScrollingTreeNode {
     
     void scrollbarVisibilityDidChange(ScrollbarOrientation, bool);
     void scrollbarMinimumThumbLengthDidChange(ScrollbarOrientation, int);
+    void markHasScrollbarStyleChange();
```
Unused.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/Scrollbar.cpp:
```diff
@@ -539,4 +539,13 @@ int Scrollbar::minimumThumbLength() const
+{
+    if (!isCustomScrollbar()) {
+        int thickness = ScrollbarTheme::theme().scrollbarThickness(widthStyle());
+        setFrameRect(IntRect(0, 0, thickness, thickness));
```
This seems wrong (since the scrollbar may have been resized since creation). You can set both height and width to the thickness. If this is a real bug, it should be covered by a test.

nmoucht replied:
Not sure what you mean here. This is following the existing code in `Scrollbar`. This comment seems relevant though: 
>   // FIXME: This is ugly and would not be necessary if we fix cross-platform code to actually query for
>     // scrollbar thickness and use it when sizing scrollbars (rather than leaving one dimension of the scrollbar
>     // alone when sizing).

smfr replied:
It would be clearer to only touch the thickness axis.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
```diff
@@ -56,6 +56,10 @@ class RemoteScrollbarsController final : public WebCore::ScrollbarsController {
 
+    void updateScrollerStyle() final;
+
+    void updateScrollerStyleForScrollableAreas(const WeakHashSet<ScrollableArea>*) final;
```
Adding this here makes the responsibilities of `ScrollbarsController` less clear. Up until now, its role is to manage a single pair of scrollbars for a single ScrollableArea. I think you should preserve that, and not make it see ScrollableAreas that it does not manage.

nmoucht replied:
Right removed this.

smfr commented on Source/WebCore/platform/ScrollbarsController.cpp:
```diff
@@ -67,4 +67,15 @@ void ScrollbarsController::mayBeginScrollGesture()
 
+void ScrollbarsController::updateScrollbarsThickness()
+{
+    Scrollbar* verticalScrollbar = scrollableArea().verticalScrollbar();
```
`auto`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/ScrollbarsController.h:
```diff
@@ -100,6 +101,12 @@ class ScrollbarsController {
+
+    WEBCORE_EXPORT void updateScrollbarsThickness();
+
+    WEBCORE_EXPORT virtual void updateScrollerStyleForScrollableAreas(const WeakHashSet<ScrollableArea>*) { }
```
Remove this (see below).

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/mac/ScrollbarsControllerMac.h:
```diff
@@ -90,7 +90,7 @@ class ScrollbarsControllerMac : public ScrollbarsController {
 
     // Public to be callable from Obj-C.
-    void updateScrollerStyle();
+    void updateScrollerStyle() final;
```
The whole class could be `final`

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
```diff
@@ -92,7 +92,7 @@ friend class RemoteScrollingCoordinatorProxyMac;
     std::unique_ptr<RemoteLayerTreeDisplayLinkClient> m_displayLinkClient;
     WebCore::PlatformLayerIdentifier m_pageScalingLayerID;
-    bool m_usesOverlayScrollbars { false };
+    std::optional<bool> m_usesOverlayScrollbars { };
```
We try to avoid `std::optional<bool>` because it can get very confusing. It's better to write the code in a way where we can always have a value.

nmoucht replied:
Changed to use ScrollerStyle enum. I don't think there's a good way to implement this without using an optional.

smfr commented on Source/WebCore/platform/Scrollbar.cpp:
```diff
@@ -547,4 +547,13 @@ int Scrollbar::minimumThumbLength() const
+{
+    if (!isCustomScrollbar()) {
+        int thickness = ScrollbarTheme::theme().scrollbarThickness(widthStyle());
+        setFrameRect(IntRect(0, 0, thickness, 0));
```
Do we expect a layout to compute the final frame rect?

nmoucht replied:
Yeah exactly. We force a layout via `scrollableArea().scrollbarStyleChanged()`.

smfr commented on Source/WebCore/platform/ScrollbarsController.cpp:
```diff
@@ -67,4 +67,15 @@ void ScrollbarsController::mayBeginScrollGesture()
+void ScrollbarsController::updateScrollbarsThickness()
+{
+    auto verticalScrollbar = scrollableArea().verticalScrollbar();
+    if (verticalScrollbar)
```
`if (auto verticalScrollbar = scrollableArea().verticalScrollbar())...`

nmoucht replied:
Fixed.

smfr commented on LayoutTests/fast/scrolling/mac/scrollable-area-size-for-overlay-scrollbars.html:
```diff
@@ -0,0 +1,43 @@
+        description("Tests that scrollable area size is properly updated when toggling scrollbar style");
+        window.internals.setUsesOverlayScrollbars(true);
+
+        var scroller;
```
Unused?

nmoucht replied:
Removed.

smfr commented on Source/WebCore/testing/Internals.h:
```diff
@@ -409,6 +409,7 @@ class Internals final : public RefCounted<Internals>, private ContextDestruction
 
     ExceptionOr<unsigned> wheelEventHandlerCount();
     ExceptionOr<unsigned> touchEventHandlerCount();
+    ExceptionOr<unsigned> scrollableAreaSize(Node&);
```
"area size" is ambiguous. Is it a size, or an area? Returning an area doesn't seem like a great way to tell if the size changed (height and width could swap, for example).

nmoucht replied:
Changed this to return just the width.

smfr commented on LayoutTests/fast/scrolling/mac/scrollable-area-size-for-overlay-scrollbars.html:
```diff
@@ -0,0 +1,43 @@
+            preChangeSize = window.internals.scrollableAreaSize(document.getElementById("scroller"));
+            window.internals.setUsesOverlayScrollbars(false);
+            await UIHelper.ensurePresentationUpdate();
+            shouldBeGreaterThan("preChangeSize", "window.internals.scrollableAreaSize(document.getElementById(\"scroller\"))");
```
I think you could do this without new Internals API. You could just have a `width=100%` element inside the scrollbar and get its `offsetWidth`.

nmoucht replied:
So the original test actually didn't properly expose the bug as the prior code did properly change the size of the scrollable area. The only way I found to expose the bug was via getting at the actual contents size of the scrollable area, which I think we can only do via the internals api.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
```diff
@@ -101,6 +101,8 @@ friend class RemoteScrollingCoordinatorProxyMac;
 
     bool m_usesOverlayScrollbars { false };
 
+    std::optional<WebCore::ScrollbarStyle> m_scrollbarStyle { };
```
No need for `{ }`

nmoucht replied:
Fixed.

nmoucht committed:
REGRESSION (UI-side compositing) changing between non-overlay and overlay scrollbars should adjust the size of the scrollable area
https://bugs.webkit.org/show_bug.cgi?id=263618
rdar://117507268

Reviewed by Simon Fraser.

This patch fixes a couple bugs related to switching back and forth dynamically between
overlay and non-overlay scrollbars. First, setFrameScrollingNodeState() was not always
called when switching the default, so iterate through each page and call
LocalFrameView::overlayScrollbarDefaultDidChange to force a scrolling commit to plumb
across this state change. Second, when we get the state change on the ui-process side,
we call _updateAllScrollerImpPairsForNewRecommendedScrollerStyle, which notifies all the
scroller imp pairs in the process of the new style. On the web process side, we had code
to adjust the scrollbar thickness and force a relayout, which now doesn't work since there
are no NSScrollerImpPairs in the web process. Instead, replicate this code in
RemoteScrollbarsController and loop through the scrollable areas in the FrameView. This
takes care of updating scrollable areas with overflow, while
https://github.com/WebKit/WebKit/pull/19561 will take care of updating scrollable areas
without overflow. Also add some additional testing infrastructure to compare the size of
the scrollable area before and after toggling scrollbar style.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::usesOverlayScrollbars const):
(WebCore::LocalFrameView::overlayScrollbarDefaultDidChange):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setFrameScrollingNodeState):
* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::scrollingTreeNodeMarkScrollbarStyleChange):
(WebCore::ScrollingTree::scrollingTreeNodesWithScrollbarStyleChange):
(WebCore::ScrollingTree::clearScrollingTreeNodesWithScrollbarStyleChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::markHasScrollbarStyleChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(-[WebScrollerImpPairDelegateMac scrollerImpPair:updateScrollerStyleForNewRecommendedScrollerStyle:]):
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::updateScrollbarThickness):
* Source/WebCore/platform/Scrollbar.h:
* Source/WebCore/platform/ScrollbarsController.cpp:
(WebCore::ScrollbarsController::updateScrollbarsThickness):
* Source/WebCore/platform/ScrollbarsController.h:
(WebCore::ScrollbarsController::updateScrollerStyle):
* Source/WebCore/platform/mac/ScrollbarsControllerMac.h:
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingNodesHaveScrollbarStyleChange):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeMarkScrollbarStyleChange):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::macScrollbarTheme):
(WebKit::RemoteScrollbarsController::updateScrollerStyle):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.messages.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::scrollingTreeNodesHaveScrollbarsWithStyleChange):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::scrollerStylePreferenceChanged):

Canonical link: https://commits.webkit.org/276439@main
